exporter/signalfxexporter/exporter.go

```diff
@@ -0,0 +1,213 @@
+	"go.uber.org/zap"
+)
+
+// New returns a new SignalFx exporter.
+func New(
+	config *Config,
+	logger *zap.Logger,
+) (exporter.MetricsExporter, error) {
+
+	if config == nil {
+		return nil, errors.New("nil config")
+	}
+
+	if config.Name() == "" {
+		config.SetType(typeStr)
+		config.SetName(typeStr)
+	}
+
+	if config.Realm == "" {
+		err := fmt.Errorf(
+			"%q config requires a non-empty \"realm\"",
+			config.Name())
+		return nil, err
+	}
+
+	if config.Timeout == 0 {
+		config.Timeout = 5 * time.Second
+	}
+
+	if config.Timeout < 0 {
+		err := fmt.Errorf(
+			"%q config cannot have a negative \"timeout\"",
+			config.Name())
+		return nil, err
+	}
+
+	headers, err := buildHeaders(config)
+	if err != nil {
+		return nil, err
+	}
+
+	url := fmt.Sprintf("https://ingest.%s.signalfx.com/v2/datapoint", config.Realm)
+
+	s := &httpSender{
+		url:     url,
+		headers: headers,
+		client: &http.Client{
+			// TODO: What other settings of http.Client to expose via config?
+			//  Or what others change from default values?
+			Timeout: config.Timeout,
+		},
+		logger: logger,
+		zippers: sync.Pool{New: func() interface{} {
+			return gzip.NewWriter(nil)
+		}},
+	}
+
+	exp, err := exporterhelper.NewMetricsExporter(
+		&config.ExporterSettings,
+		s.pushMetricsData,
+		exporterhelper.WithTracing(true),
+		exporterhelper.WithMetrics(true))
+
+	return exp, err
+}
+
+// httpSender sends the data to the SignalFx backend.
+type httpSender struct {
+	url     string
+	headers map[string]string
+	client  *http.Client
+	logger  *zap.Logger
+	zippers sync.Pool
+}
+
+func (s *httpSender) pushMetricsData(
+	ctx context.Context,
+	md consumerdata.MetricsData,
+) (droppedTimeSeries int, err error) {
+
+	sfxDataPoints, numDroppedTimeseries, err := metricDataToSingalFxV2(s.logger, md)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+	}
+
+	body, compressed, err := s.encodeBody(sfxDataPoints)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+	}
+
+	req, err := http.NewRequest("POST", s.url, body)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+	}
+
+	if s.headers != nil {
+		for k, v := range s.headers {
+			req.Header.Set(k, v)
+		}
+	}
+
+	if compressed {
+		req.Header.Set("Content-Encoding", "gzip")
+	}
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), err
+	}
+
+	io.Copy(ioutil.Discard, resp.Body)
+	resp.Body.Close()
+
+	// SignalFx accepts all 2XX codes.
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
+		err = fmt.Errorf(
+			"HTTP %d %q",
+			resp.StatusCode,
+			http.StatusText(resp.StatusCode))
+		return exporterhelper.NumTimeSeries(md), err
+	}
+
+	return numDroppedTimeseries, nil
+}
+
+func buildHeaders(config *Config) (map[string]string, error) {
+	if config.AuthToken == "" {
+		err := fmt.Errorf(
```

Comment:
There are scenarios where the auth token is not required, such as if the data is going to another proxy that provides its own access token.  That might not be in the immediately roadmap for our intended deployment models, but I would not impose this requirement on this exporter.

Reply:
I have had to go back and do patch releases of various integrations because I assumed the access token would always be required and it turned out later it wasn't.